README.rst
============
 Eve-Mocker
============

Mocking tool for `Eve powered REST API <http://python-eve.org>`_, based on the excellent `HTTPretty <http://falcao.it/HTTPretty>`_, aimed to be used in your unit tests, when you rely on an Eve API.

Features
========

Eve-Mocker doesn't try to replicate every Eve features, by design, it doesn't need any Eve settings files, doesn't support schema validation and more advanced features. Don't hesitate to contribute if you need more complex features.

* Everything is stored in memory (``self.items``)
* Support all methods except HEAD requests
* Handle ETags, and always return meaningful status code, like Eve.
* Partial support of filtering and sorting (only mongo query syntax)
* No need to change your code for testing, HTTPretty does everything for you, it works well with `requests <http://www.python-requests.org>`_.
* It renders only JSON, no XML yet.

Installing
==========

(soon to be released on pypy.)

.. code-block::

    $ pip install eve-mocker


Usage
=====

You should read `HTTPretty <http://falcao.it/HTTPretty>`_ documentation before starting.

Here is a basic example:

.. code-block:: python

    from httpretty import HTTPretty
    import requests

    HTTPretty.enable()
    EveMocker("http://myapi.com/api/")
    
    response = requests.get("http://myapi.com/api/mymodel/")
    assert response.status_code == 200
    assert response.json() == {"_items": []}

    HTTPretty.disable()

You can add data directly with ``EveMocker.set_resource``:

.. code-block:: python

    from httpretty import HTTPretty
    import requests

    HTTPretty.enable()
    eve_mocker = EveMocker("http://myapi.com/api/")

    mymodel_data = [{"_id": "fakeid", "content": "mycontent"}]
    eve_mocker.set_resource("mymodel", mymodel_data)

    response = requests.get("http://myapi.com/api/mymodel/")
    assert response.status_code == 200
    assert response.json() == {"_items": mymodel_data}

    HTTPretty.disable()


Alternatively, you use ``EveMocker`` within a context manager, and it will automatically call ``HTTPretty.enable()`` and ``HTTPretty.disable()``.

.. code-block:: python

    from eve_mocker import EveMocker
    import requests

    with EveMocker("http://myapi.com/api/"):
        response = requests.get("http://myapi.com/api/mymodel/")
        assert response.status_code == 200
        assert response.json() == {"_items": []}

Check the tests (in ``test_eve_mocker.py``) if you want to check more examples.

Getting started
===============

Let's say you want to test the following class stored in ``remote_items.py`` that need to call an Eve powered REST API:

.. code-block:: python

    from urlparse import urljoin
    from functools import partial
    import requests

    API_URL = "http://my-eve-api.com/api/"


    class RemoteItems(object):
        def __init__(self, api_url=API_URL):
            self.api_url = api_url
            self.endpoint_url = partial(urljoin, self.api_url)

        def get_latest(self):
            r = requests.get(self.endpoint_url("items"))
            r.raise_for_status()
            return r.json().get("_items", [])


Here is how you can do it with Eve-Mocker:

.. code-block:: python

    from eve_mocker import EveMocker
    import unittest
    from remote_items import RemoteItems, API_URL

    class TestRemoteItems(unittest.TestCase):
        def testGetLatestItems(self):
            items = [{"_id": "fakeid", "content": "my content"},
                     {"_id": "fakeid2", "content": "another_content"}]
            with EveMocker(API_URL) as eve_mocker:
                # We feed EveMocker DB with some items
                eve_mocker.set_resource("items", items)

                remote_items = RemoteItems()
                latest_items = remote_items.get_latest()

                self.assertEqual(sorted(latest_items), sorted(items))

    if __name__ == '__main__':
        unittest.main()


You can find these two files in the **examples** directory.


Advanced Usage
==============

``EveMocker`` takes two additonals arguments, ``default_pk`` if you need a primary key other than ``_id``, and ``pk_maps`` which is a mapping resource => primary key: ``{"resource": "pk_field", "resource2": "pk_field"}``.

License (MIT)
=============

Copyright (c) 2013 Thomas Sileo

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGES.rst
======================
 Eve-Mocker Changelog
======================

0.1.0 (2013-06-24)
==================

- First version.


test_eve_mocker.py
# -*- coding: utf-8 -*-

""" test_eve_mocker.py - Test the eve_mocker module. """

import unittest
import requests
from httpretty import HTTPretty
from sure import expect
from eve_mocker import EveMocker, query_data
from urlparse import urljoin
from functools import partial
import json

BASE_URL = "http://localhost/api/"
api_url = partial(urljoin, BASE_URL)


class TestEveMocker(unittest.TestCase):
    def setUp(self):
        HTTPretty.enable()
        self.eve_mocker = EveMocker(BASE_URL, default_pk="testpk")

    def tearDown(self):
        HTTPretty.disable()

    def testAPI(self):
        """ Testing all client features. """
        mymodel_url = api_url("mymodel")
        mymodel1 = {"testpk": "mypk1", "content": "test content"}

        # Check that mymodel resource is empty
        response = requests.get(mymodel_url)
        data = response.json()

        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("_items")
        expect(data["_items"]).to.be.empty

        # Posting mymodel1
        response = requests.post(mymodel_url,
                                 {"mymodel1": json.dumps(mymodel1)})
        data = response.json()

        # Check the status of the item and if it has an etag
        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("mymodel1")
        expect(data["mymodel1"]["status"]).to.equal("OK")
        expect(data["mymodel1"]).to.have.key("etag")

        # Storing the ETag for later
        mymodel1_etag = data["mymodel1"]["etag"]

        # Check that it has actually been created
        response = requests.get(mymodel_url)
        mymodel1_test = mymodel1.copy()
        mymodel1_test.update({"etag": data["mymodel1"]["etag"]})

        expect(response.status_code).to.equal(200)
        expect(response.json()).to.equal({"_items": [mymodel1_test]})

        # Check if we can retrieve the item via its URI
        response = requests.get(api_url("mymodel/mypk1/"))
        data = response.json()

        expect(response.status_code).to.equal(200)

        expect(data).to.equal(mymodel1_test)

        # Check that we CAN'T rePOST mymodel1 with the same primary key
        response = requests.post(mymodel_url,
                                 {"mymodel1": json.dumps(mymodel1)})
        data = response.json()

        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("mymodel1")
        expect(data["mymodel1"]["status"]).to.equal("ERR")

        # Now we try to PATCH the item without If-Match header
        mymodel1_patch = {"content": "new content"}
        response = requests.patch(api_url("mymodel/mypk1/"),
                                  {"data": json.dumps(mymodel1_patch)})

        expect(response.status_code).to.equal(403)

        # Check that it doesn't work with the wrong ETag
        response = requests.patch(api_url("mymodel/mypk1/"),
                                  {"data": json.dumps(mymodel1_patch)},
                                  headers={"If-Match": "falsyetag"})

        expect(response.status_code).to.equal(412)

        # Finally we PATCH with the good ETag
        response = requests.patch(api_url("mymodel/mypk1/"),
                                  {"data": json.dumps(mymodel1_patch)},
                                  headers={"If-Match": mymodel1_etag})
        data = response.json()

        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("data")
        expect(data["data"]["status"]).to.equal("OK")
        expect(data["data"]).to.have.key("etag")

        mymodel1_etag = data["data"]["etag"]
        mymodel1_test.update({"etag": mymodel1_etag,
                              "content": "new content"})

        # Check if the item has been updated
        response = requests.get(api_url("mymodel/mypk1/"))
        data = response.json()

        expect(response.status_code).to.equal(200)
        expect(data).to.equal(mymodel1_test)

        # Delete without ETag should return 403
        response = requests.delete(api_url("mymodel/mypk1/"))

        expect(response.status_code).to.equal(403)

        # Delete with a WRONG ETag should return 412
        response = requests.delete(api_url("mymodel/mypk1/"),
                                   headers={"If-Match": "wrongetag"})

        expect(response.status_code).to.equal(412)

        # Now we delete it the right way
        response = requests.delete(api_url("mymodel/mypk1/"),
                                   headers={"If-Match": mymodel1_etag})

        expect(response.status_code).to.equal(200)

        # Check that mymodel resource is empty
        response = requests.get(mymodel_url)
        data = response.json()

        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("_items")
        expect(data["_items"]).to.be.empty

    def testSetResource(self):
        """ Test that EveMocker.set_resource works. """
        test_items = sorted([{"testpk": "pk1", "content": "content1"},
                             {"testpk": "pk2", "content": "content2"}])

        self.eve_mocker.set_resource("testresource", test_items)

        # Check that we retireve the items on a GET call
        response = requests.get(api_url("testresource/"))
        data = response.json()

        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("_items")
        expect(sorted(data["_items"])).to.equal(test_items)

    def testFiltering(self):
        """ Test ?where query with basic filtering. """
        test_items = sorted([{"testpk": "pk1", "content": "content1"},
                             {"testpk": "pk2", "content": "content2"}])

        self.eve_mocker.set_resource("testresource", test_items)

        # Check that we retireve the items on a GET call
        response = requests.get(api_url('testresource/?where={"content": "content1"}'))
        data = response.json()

        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("_items")
        expect(sorted(data["_items"])).to.equal([{"testpk": "pk1", "content": "content1"}])

    def testFilteringMongoQuery(self):
        """ Test ?where query with with mongo query syntax filtering. """
        test_items = [{"testpk": i} for i in range(50)]

        self.eve_mocker.set_resource("testresource", test_items)

        # Check that we retireve the items on a GET call
        response = requests.get(api_url('testresource/?where={"testpk": {"$gte": 10}}'))
        data = response.json()

        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("_items")
        expect(sorted(data["_items"])).to.have.length_of(40)

    def testSort(self):
        """ Test ?sort query with with mongo syntax. """
        test_items = [{"testpk": i} for i in range(50)]

        self.eve_mocker.set_resource("testresource", test_items)

        # Check that we retireve the items on a GET call
        response = requests.get(api_url('testresource/?sort={"testpk": 1}'))
        data = response.json()

        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("_items")
        expect(data["_items"][0]).to.equal({"testpk": 0})

    def testSortDesc(self):
        test_items = [{"testpk": i} for i in range(50)]

        self.eve_mocker.set_resource("testresource", test_items)

        response = requests.get(api_url('testresource/?sort={"testpk": -1}'))
        data = response.json()
        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("_items")
        expect(data["_items"][0]).to.equal({"testpk": 49})

    def testSortDescWithWhere(self):
        test_items = [{"testpk": i} for i in range(50)]

        self.eve_mocker.set_resource("testresource", test_items)

        response = requests.get(api_url('testresource/?sort={"testpk": -1}&where={"testpk": {"$lte": 30}}'))
        data = response.json()
        expect(response.status_code).to.equal(200)
        expect(data).to.have.key("_items")
        expect(data["_items"][0]).to.equal({"testpk": 30})

    def testSetResourceNoPk(self):
        """ Set a resource item without PK should raise an Exception. """
        # No pk for the item should raise an Exception
        falsy_items = [{"content": "content1"}]
        set_resource = self.eve_mocker.set_resource

        expect(set_resource).when.called_with("testresource",
                                              falsy_items).should.throw(Exception)

    def testQueryData(self):
        """ Test if mongo query syntax works, testing the raw query_data """
        test_data = [{"testpk": i} for i in range(50)]
        res = query_data(test_data, {"testpk": 10})

        expect(res).to.have.length_of(1)
        expect(res[0]).to.equal(test_data[10])

        res = query_data(test_data, {"testpk": {"$gte": 10}})
        expect(res).to.have.length_of(40)

        res = query_data(test_data, {"testpk": {"$gte": 10, "$lte": 15}})
        expect(res).to.have.length_of(6)

        res = query_data(test_data, {"badkey": {"$gte": 10, "$lte": 15}})
        expect(res).to.have.length_of(0)

        res = query_data(test_data, {"testpk": {"$in": [1, 2, 3, 4, 5]}})
        expect(res).to.have.length_of(5)

        res = query_data(test_data, {"testpk": {"$nin": [1, 2, 3, 4, 5]}})
        expect(res).to.have.length_of(45)

        res = query_data(test_data, {"testpk": {"$ne": 10}})
        expect(res).to.have.length_of(49)

        res = query_data(test_data, {"testpk": {"$gt": 10}})
        expect(res).to.have.length_of(39)

        res = query_data(test_data, {"testpk": {"$lt": 10}})
        expect(res).to.have.length_of(10)

    def testContextManager(self):
        """ Test EveMocker within a context manager. """
        with EveMocker("http://myapi.com/api/"):
            response = requests.get("http://myapi.com/api/mymodel")
            expect(response.status_code).to.equal(200)
            expect(response.json()).to.equal({"_items": []})


if __name__ == '__main__':
    unittest.main()


examples/test_remote_items.py
from eve_mocker import EveMocker
import unittest
from remote_items import RemoteItems, API_URL


class TestRemoteItems(unittest.TestCase):
    def testGetLatestItems(self):
        items = [{"_id": "fakeid", "content": "my content"},
                 {"_id": "fakeid2", "content": "another_content"}]
        with EveMocker(API_URL) as eve_mocker:
            # We feed EveMocker DB with some items
            eve_mocker.set_resource("items", items)

            remote_items = RemoteItems()
            latest_items = remote_items.get_latest()

            self.assertEqual(sorted(latest_items), sorted(items))

if __name__ == '__main__':
    unittest.main()


eve_mocker.py
# -*- coding: utf-8 -*-
import re
import json
from httpretty import HTTPretty
from urlparse import parse_qs, urljoin
from collections import defaultdict
import time
import hashlib
from itertools import ifilter

PK = "_id"


def generate_etag():
    """ Helper function for generating random etag. """
    return hashlib.sha1(str(time.time())).hexdigest()


def query_data(data, q):
    """ Filter data for a given MongoDB query.

    Only a subset of MongoDB query language is supported:

    - $gt
    - $gte
    - $lt
    - $lte
    - $ne
    - $in
    - $nin

    You can combine multiple operator like {"val": {"$gt": 10, "$lt": 20}}

    :type data: list
    :param data: List of dict

    :type q: dict
    :param q: MongoDB Query

    """
    for k, f in q.items():
        if isinstance(f, dict):
            for _filter, val in f.items():
                _f = lambda x: True
                if _filter.startswith("$"):
                    if _filter == "$gt":
                        _f = lambda x: k in x and x[k] > val
                    elif _filter == "$gte":
                        _f = lambda x: k in x and x[k] >= val
                    elif _filter == "$lt":
                        _f = lambda x: k in x and x[k] < val
                    elif _filter == "$lte":
                        _f = lambda x: k in x and x[k] <= val
                    elif _filter == "$ne":
                        _f = lambda x: k in x and x[k] != val
                    elif _filter == "$in":
                        _f = lambda x: k in x and x[k] in val
                    elif _filter == "$nin":
                        _f = lambda x: k in x and not x[k] in val
                else:
                    _f = lambda x: k in x and x[k] == f
                data = list(ifilter(_f, data))
        else:
            _f = lambda x: k in x and x[k] == f
            data = list(ifilter(_f, data))
    return data


class EveMocker(object):
    """ Eve API mocker

    Everything is stored in self.items.

    Support:
        - All methods except HEAD
        - Handle etags
        - Return good status_code

    Not (yet?) supported:
        - HATEOAS links
        - schema validation

    :type base_url: str
    :param base_url: API base url

    :type pk_maps: dict
    :param pk_maps: Primary key mapping {resource: pk}

    :type default_pk: str
    :param default_pk: Default primary key, _id by default.

    """
    def __init__(self, base_url, pk_maps={}, default_pk="_id"):
        self.items = defaultdict(dict)
        self.base_url = base_url
        self.default_pk = default_pk
        self.pk_maps = pk_maps
        self.etag = {}

        # Register all URIs for resources
        resource_url = urljoin(self.base_url, "([^/]+/?$)")
        resource_regex = re.compile(resource_url)

        HTTPretty.register_uri(HTTPretty.GET,
                               resource_regex,
                               body=self.generate_resource_response)
        HTTPretty.register_uri(HTTPretty.POST,
                               resource_regex,
                               body=self.generate_resource_response)

        HTTPretty.register_uri(HTTPretty.DELETE,
                               resource_regex,
                               body=self.generate_resource_response)

        # Register URIs for items
        item_regex = re.compile(urljoin(self.base_url, "([a-zA-Z0-9-_]+)/([a-zA-Z0-9-_]+)"))

        HTTPretty.register_uri(HTTPretty.GET,
                               item_regex,
                               body=self.generate_item_response)

        HTTPretty.register_uri(HTTPretty.PATCH,
                               item_regex,
                               body=self.generate_item_response)

        HTTPretty.register_uri(HTTPretty.DELETE,
                               item_regex,
                               body=self.generate_item_response)

    def __enter__(self):
        HTTPretty.enable()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        HTTPretty.disable()

    def get_pk(self, resource):
        return self.pk_maps.get(resource, self.default_pk)

    def get_resource(self, resource):
        """ Retrieve all items for a resource.

        Resources are stored in a dict of list (self.items).

        :type resource: str
        :param resource: Resource name

        """
        return [item for key, item in self.items[resource].items()]

    def set_resource(self, resource, items=[]):
        """ Set items for a given resource.

        :type resource: str
        :param resource: Resource name

        :type items: list
        :param items: List of items (dict),
            also don't forget the pk for each items.

        """
        pk = self.get_pk(resource)
        for item in items:
            if pk in item:
                self.items[resource][item[pk]] = item
            else:
                exc = "No primary key: {0} found for item: {1}".format(pk, item)
                raise Exception(exc)

    def generate_resource_response(self, request, uri, headers):
        """ Generate a response for a resource,
            support all methods, except HEAD. """
        headers["content_type"] = "application/json"
        path = filter(lambda x: x not in ["api", ""], request.path.split('/'))
        resource = path[0]
        if request.method == "GET":
            _items = self.get_resource(resource)

            # Check if a querystring is provided
            if request.querystring and "where" in request.querystring:
                try:
                    # Load the mongo query and filter the result
                    q = json.loads(request.querystring["where"][0])
                    _items = query_data(_items, q)
                except:
                    pass
            if request.querystring and "sort" in request.querystring:
                try:
                    s = json.loads(request.querystring["sort"][0])
                    for k, v in s.items():
                        reverse = True
                        if v == 1:
                            reverse = False
                        elif v == -1:
                            reverse = True
                        _items = sorted(_items, key=lambda x: x[k], reverse=reverse)
                except:
                    pass

            return [200,
                    headers,
                    json.dumps({"_items": _items})]

        elif request.method == "POST":
            qs = parse_qs(request.body)
            out = {}
            for key, data in qs.items():
                item = json.loads(data[0])
                pk = item[self.get_pk(resource)]
                if pk in self.items[resource]:
                    out[key] = {"status": "ERR", "issues": ["pk not unique"]}
                else:
                    item_etag = generate_etag()
                    item["etag"] = item_etag
                    self.items[resource][pk] = item
                    out[key] = {"status": "OK", "etag": item_etag}
            return [200, headers, json.dumps(out)]

        elif request.method == "DELETE":
            del self.items[resource]
            return [200, headers, "{}"]

    def generate_item_response(self, request, uri, headers):
        """ Generate a response for an item,
            support all methods, except HEAD. """
        headers["content_type"] = "application/json"
        resource, item_id = filter(lambda x: x not in ["api", ""],
                                   request.path.split('/'))
        if item_id not in self.items[resource]:
            status_code = 405
            if request.method == "GET":
                status_code = 404
            return [status_code, headers, "{}"]

        if request.method in ["PATCH", "DELETE"]:
            if not "If-Match" in request.headers:
                return [403, headers, "{}"]

            old_etag = self.items[resource][item_id]["etag"]

            if request.headers["If-Match"] != old_etag:
                return [412, headers, "{}"]

        if request.method == "GET":
            r = self.items[resource]
            if item_id in r:
                return [200, headers, json.dumps(r[item_id])]

        elif request.method == "DELETE":
            r = self.items[resource]
            if item_id in r:
                del self.items[resource][item_id]
                return [200, headers, "{}"]

        elif request.method == "PATCH":
            qs = parse_qs(request.body)
            out = {}
            for k, patch_data in qs.items():
                patch_data = json.loads(patch_data[0])
                new_etag = generate_etag()
                patch_data.update({"etag": new_etag})
                self.items[resource][item_id].update(patch_data)
                out[k] = {"status": "OK", "etag": new_etag}
            return [200, headers, json.dumps(out)]

        return [405, headers, "{}"]
